### DISCUSSION TITLE
SDL3 backend support :)

### DISCUSSION THREAD

**Discussion by ghost** [pthom/hello_imgui#129] [Ideas]
It has a lot of new features...
https://wiki.libsdl.org/SDL3/NewFeatures
so it will be nice to have *hello_imgui* ready for.

**Comment by pthom** [comment]
Hello,

Support for SDL3 is not currently being worked on, as I’m juggling multiple projects with limited time.

For information here are some more details if someone want to contribute on this.

Adding a new platform backend requires significant care and effort because Hello ImGui needs to remain compatible with various combinations of:
1. **Platform backends**: SDL, GLFW  
2. **Rendering backends**: OpenGL, Metal, DirectX (11/12), Vulkan  
3. **Operating systems**: Linux, macOS, Windows, iOS, Android, and Emscripten  

All these combinations must be manually tested across CI pipelines.

Additionally, introducing a new backend would need to maintain core features such as:
- Consistent window size and position handling across platforms
- Support for fullscreen apps and monitor selection
- Proper high-DPI handling
- Application icon configuration  

### Areas Requiring Changes for SDL3 Support:

#### C++
- `src/hello_imgui/internal/backend_impls/runner_sdl2.[h-cpp]`: Main implementation
- `src/hello_imgui/internal/backend_impls/backend_window_helper/sdl_window_helper.[h-cpp]`: Window creation and fullscreen handling
- `src/hello_imgui/internal/backend_impls/opengl_setup_helper/opengl_setup_sdl.[h-cpp]`: OpenGL-specific setup
- `src/hello_imgui/internal/backend_impls/runner_sdl_emscripten.h`: Emscripten-related handling
- `src/hello_imgui/internal/backend_impls/win32_dpi_awareness`: High-DPI handling

#### CMake
- `hello_imgui_cmake/hello_imgui_build_lib.cmake`: Central location for all **HelloImGui**'s CMake options  

If/when SDL3 support is considered, careful planning and execution will be necessary to ensure:
- compatibility and feature parity
- testing and maintenance accross all combination of OS / platform backend / rendering backend

### Currently Supported combinations of OS / Platform Backend / Rendering backend 

There are currently **21** supported combinations (which is a lot to maintain, and test whether manually and in CI).
Adding support for SDL3 would mechanically increase this number by a factor of approx 1.5.

| Operating System | Platform Backend | Rendering Backend         |
|------------------|-------------------|---------------------------|
| **Linux**        | GLFW3            | OpenGL                    |
| **Linux**        | SDL2             | OpenGL                    |
| **Linux**        | GLFW3            | Vulkan                    |
| **Linux**        | SDL2             | Vulkan                    |
| **macOS**        | GLFW3            | OpenGL                    |
| **macOS**        | SDL2             | OpenGL                    |
| **macOS**        | GLFW3            | Metal                     |
| **macOS**        | SDL2             | Metal                     |
| **Windows**      | GLFW3            | OpenGL                    |
| **Windows**      | SDL2             | OpenGL                    |
| **Windows**      | GLFW3            | Vulkan                    |
| **Windows**      | SDL2             | Vulkan                    |
| **Windows**      | GLFW3            | DirectX 11                |
| **Windows**      | SDL2             | DirectX 11                |
| **Windows**      | GLFW3            | DirectX 12                |
| **Windows**      | SDL2             | DirectX 12                |
| **iOS**          | SDL2             | OpenGL                    |
| **iOS**          | SDL2             | Metal                     |
| **Android**      | SDL2             | OpenGL                    |
| **Android**      | SDL2             | Vulkan                    |
| **Emscripten**   | SDL2             | OpenGL (WebGL)            |
